### PR TITLE
Add sdist to release uploads

### PR DESCRIPTION
**Pull request by dbast**
#13663 added the workflow to upload a reproducible release tarball to the release artifacts to not rely on the non-stable Github mechanisms that provide those.

While archiving the entire repo via `git archive` is nice, the real Pythonic artifact is a sdist (Source Distribution) file as also packages upload to pypi.

This PR adds the sdist creation to the upload flow and also changes the sha256sum calculation to produce a combined SHA256SUMS file, as the sdist file name is not something we influence and can rely on in the flow, but is given by hatch and the configuration in the pyproject.toml. So a tagged releases produces a conda-$tag.tar.gz (size 550 KB, name produced by hatch) and conda-$tag_full.tar.gz (size 21 MB, name defined in the flow) file and a combined SHA256SUMS file (huge size difference!). The sdist produced by hatch is also reproducible (according to the hatch documentation).

The sdist file conda-$tag.tar.gz should be complete by concept and should be enough as source file for the feedstocks to build conda and it is the first step, if we ever upload conda to pypi (after all the dependencies are available there). 

(The build command can also easily produce a wheel additionally by adding the `--wheel` arg: `python -m build --sdist --wheel --outdir release`)

### Checklist - did you ...

- [ ] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [ ] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

File: .github/workflows/upload.yml [hunk 1]
```diff
@@ -37,13 +37,24 @@ jobs:
...
+      - name: Install build
+        run: pip install build
+
+      - name: Build Source Distribution
+        run: python -m build --sdist --outdir release
+
+      - name: Compute Checksums
+        run: |
+          cd release
+          sha256sum * > SHA256SUMS
```
Comment (jezdez) [inline .github/workflows/upload.yml:57]: Is SHA256SUMS a common name for that?

File: .github/workflows/upload.yml [hunk 2]
```diff
@@ -37,13 +37,24 @@ jobs:
       - name: Archive Source
         run: git archive
           --prefix="${{ env.ARCHIVE_NAME }}/"
-          --output="release/${{ env.ARCHIVE_NAME }}.tar.gz"
+          --output="release/${{ env.ARCHIVE_NAME }}_full.tar.gz"
```
Comment (jezdez) [inline .github/workflows/upload.yml:40]:
```suggestion
          --output="release/${{ env.ARCHIVE_NAME }}_archive.tar.gz"
```

**Review by jezdez — COMMENTED** [review]
Given that we haven't solved https://github.com/conda/conda/issues/12126 I'm inclined to not do this, it could give users the idea that installing conda via an sdist and the Python Packaging stack is supported. It's not supported at the moment.

File: .github/workflows/upload.yml (hunk 1 above)
Comment (dbast) [reply .github/workflows/upload.yml:57]: Didn't find a standard yet, but it is e.g. used by Debian since ages, see e.g. https://cdimage.debian.org/debian-cd/current/amd64/iso-cd/SHA256SUMS

**Comment by dbast** [comment]
> Given that we haven't solved #12126 I'm inclined to not do this, it could give users the idea that installing conda via an sdist and the Python Packaging stack is supported. It's not supported at the moment.

Well, I see it more like a consistency check (not a statement that conda is installed via sdist):
* `python -m build --sdist .` runs hatch to create the sdist tarball 
* conda-build runs `{{ PYTHON }} -m pip install . --no-deps --no-build-isolation -vv && {{ PYTHON }} -m conda init --install` and thus also runs hatch to do the pip install
* this is a consistency check as hatch will only install things when called via pip as it puts into the sdist via build call... the intermediate sdist allows to inspect what hatch considers as relevant to be installed / packed and see if that is correct... (had some old CI pipelines years ago, where we first created the sdist, then installed from it to ensure the sdist is complete)

**Comment by dbast** [comment]
There are two other arguments that would speak for the sdist as distribution archive:
* the sdist has the correct hatch derived version already burned-in, thus we don't rely on a correct https://github.com/conda/conda/blob/main/.git_archival.txt file + artifact name.
* providing only the sdist to downstream feedstocks is a kind of governance pattern as it enforces consistency, by preventing that arbitrary files straight from the repo can be copied into the conda package, as everything has to then be from the sdist only.

File: .github/workflows/upload.yml (hunk 1 above)
Comment (jezdez) [reply .github/workflows/upload.yml:57]: Ah, that's where I saw it, thanks for the pointer. Yeah, it's as good as any, and some precedence is useful.

**Comment by zklaus** [comment]
Just stumbled upon this PR in a discussion with Tania. The context is that `conda` is a dependency for `conda-store`, but commented out in the `pyproject.toml` because it's not on Pypi. My personal opinion from past experience in the climate science world is that Python distribution packages should be published to Pypi even when they are not pip installable, in that case usually only as sdists, of course.

Among other things that enables a meaningful `pip check` for downstream packages and allows them to declare complete dependencies in their respective `pyproject.toml` files.

**Comment by jezdez** [comment]
> Just stumbled upon this PR in a discussion with Tania. The context is that `conda` is a dependency for `conda-store`, but commented out in the `pyproject.toml` because it's not on Pypi. My personal opinion from past experience in the climate science world is that Python distribution packages should be published to Pypi even when they are not pip installable, in that case usually only as sdists, of course.

While I appreciate your experience with Python packaging in your field of work, I'm afraid you're wrong about conda being a Python packaging tool. It's a common misunderstanding and a serious barrier of entry for Python users to adopt conda, sadly, with considerable impact on adoption. Imagine all the "regular Python users" that would benefit from conda!

Sadly pip will create wheel files out of sdists automatically and install them and there is no way to mark a package as uploaded to PyPI but unsupported to my knowledge. From my experience, once it's uploaded to PyPI, the floodgates will open and users will start to install it from there.

In other words, publishing sdists on PyPI is not appropriate before we've solved #12126 and made sure everything works as expected (especially handling the shell integration and the dependencies). This should not be done accidentally or just because it makes a release process easier.

> Among other things that enables a meaningful `pip check` for downstream packages and allows them to declare complete dependencies in their respective `pyproject.toml` files.

You're assuming that conda is currently a fully Python packaging compatible project, but we aren't there yet (maybe [PEP 725](https://peps.python.org/pep-0725/) will help). pip check will completely ignore the current way of specifying conda dependencies (packages on Anaconda's main channel or conda-forge).

**Comment by zklaus** [comment]
I understand there are many factors to consider and these decisions should not be made lightly. To clarify, I did not mean to imply that conda is "only" a Python packaging tool, merely that anything that installs a `.dist-info` directory is a [Python distribution package](https://packaging.python.org/en/latest/discussions/distribution-package-vs-import-package/) and that it is useful to have the metadata and dependency relations available from a single source like PyPI.

I have helped with the packaging of many non-Python packages for conda-forge, sometimes completely independent from Python (such as subversion), but mostly with a Python binding somewhere along the chain. In those cases `pip check` will of course not guarantee that the binary dependencies are in order, but we get the Python side covered as well as possible.

Having said all that, I hear you: Solve #12126 and everybody is happy :smiley:
